---
title: "ChIP Reanalysis"
author: "Lucas Carter"
date: "2024-10-11"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## ChIP Reanalysis for Figure Panel 6

This file contains the code necessary for generating the ChIP plots in Figure Panel 6 that use intron coverage data from total RNAseq generated for **"Extra-nucleolar Pol I regulates gene transcription through chromatin domain maintenance"**. Publicly available [ChIP data](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145874) were taken from [Genome-wide analyses of chromatin interactions after the loss of Pol I, Pol II, and Pol III](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02067-3#availability-of-data-and-materials). Mouse ESC NAD tracks were taken from [Genome-wide maps of nucleolus interactions reveal distinct layers of repressive chromatin domains](https://pubmed.ncbi.nlm.nih.gov/35304483/).  

#### Load initial packages here  

Loads initial packages and set root directory where dryad repository is stored

```{r, message=FALSE, warning=FALSE}

## Set your root directory here
root<- "/Volumes/external hd/IBiS/Backman_Lab/Transcription Publication/repositories/dryad/"

## gene position directory
bed.dir <- "/ChIP/peaks/"

require("rtracklayer")
require("ggplot2")
require("GenomicRanges")
require("IRanges")
require("plyranges")
require("TxDb.Mmusculus.UCSC.mm39.refGene")
require("BSgenome.Mmusculus.UCSC.mm39")
require("org.Mm.eg.db")
require("ChIPseeker")
require("AnnotationHub")
require("corrplot")
require("pheatmap")
require("tidyr")
require("forcats")

chroms <- c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr11", 
            "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chrX", "chrM", "chrR")

```

#### Prepare data for plotting

```{r, message=FALSE, warning=FALSE}
##------------------------------Bedtools intersected replicates------------------------------##

pol1.bed <- read.table(file.path(root,bed.dir,"pol1.bedtools.union.bed"), sep= "\t", header= TRUE)
pol2.bed <- read.table(file.path(root,bed.dir,"pol2.bedtools.union.bed"), sep= "\t", header= TRUE)

pol1.bed <- GRanges(pol1.bed[,1:10])
pol2.bed <- GRanges(pol2.bed[,1:10])

##------------------------------Annotation of reps------------------------------##
## Annotate by genomic features like promoters, exons, etc

## Prepare annotation objects for mapping
ah <- AnnotationHub()
qcall <- query(ah, c("OrgDb","Mus musculus"))
hs <- ah[[names(qcall)]]
txdb_mm39 <- TxDb.Mmusculus.UCSC.mm39.refGene

## Make TX for all genomic features, including ncRNAs and lincRNAs 
library(GenomicFeatures)
txdb <- makeTxDbFromBiomart(dataset="mmusculus_gene_ensembl") ## Grab most recent ensembl data base
tx <- transcripts(txdb, columns=c("tx_name", "gene_id", "tx_type"))

## Convert TX to Genomic Ranges data
tx <- as.data.frame(tx)
tx <- tx %>% transform( seqnames= paste0("chr",tx$seqnames), start = tx$start, end = tx$end)  %>% as_granges()


## Annotate results. Packages introduced previously
annotate.peak <- function(peaks.file){
  
  ## Call annotation
  anno <- annotatePeak(peaks.file, TxDb = tx, tssRegion=c(-3000, 3000))
  anno@anno$symbol <- mapIds(hs, anno@anno$gene_id, column = "SYMBOL", keytype = "ENSEMBL")
  anno <- data.frame(anno@anno)
  
  ## Make a counts field
  anno$Count<- rep(1, nrow(anno))
  
  return(anno)
  
}

pol1.anno <- annotate.peak(pol1.bed)
pol2.anno <- annotate.peak(pol2.bed)

#####################################################################################
# Correlation analysis of genomic features                                
#####################################################################################

## For loop to generate 1 MB segments of the genome for counting overlaps in peaks
#Bins = 10k, 25k, 100k, and 1 MB
binGenome <- function(chrom.file="mm39.chrom.sizes.csv", bin.size, 
                      bed.dir = "/ChIP/chrom_sizes/",
                      chroms = c("chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr11", 
                                 "chr12", "chr13", "chr14", "chr15", "chr16", "chr17", "chr18", "chr19", "chrX", "chrM", "chrR")){
  
  require(plyranges)
  require(dplyr)
  
  ## chrom sizes for mm10
  chrom.sizes <- read.table(file.path(root,bed.dir, chrom.file), sep= ",", header= T)
  
  ## Filter out chromosomes
  chrom.sizes <- chrom.sizes[chrom.sizes$name %in% chroms,]
  
  tbl <- data.frame()
  for (i in 1:length(chrom.sizes$name)){
    
    chrom <- chrom.sizes$name[i]

    
    if((chrom == "chrR" & bin.size > 15000 ) | ( chrom == "chrM" & bin.size > 15000)){
      
      x <-as.numeric(0)
      y <-as.numeric(chrom.sizes$size[i])
      
      name <- rep(chrom, length(x))
      
      df <- data.frame(name, x, y)

      
    } else {
      
      x <-as.numeric(seq(0,chrom.sizes$size[i], by = bin.size))
      
      y <-as.numeric(seq(bin.size,chrom.sizes$size[i], by = bin.size))
      y <- append(y, chrom.sizes$size[i], after = length(y))
      
      name <- rep(chrom, length(x))
      
      df <- data.frame(name, x, y)

      
    }
    
    tbl <- rbind(tbl,df)
    
  }
  
  ## Generate ranges from above
  tbl <- tbl %>% 
    transform( seqnames= paste0(tbl$name), start = tbl$x, end = tbl$y)  %>% 
    as_granges()
  
  return(tbl)
  
}

mm39.tbl <- binGenome("mm39.chrom.sizes.csv", 10000)
mm39.tbl

```

#### Figure 6F: Spearman correlation

```{r, message=FALSE, warning=FALSE}

#####################################################################################
# compositional analysis of NADs and rDNA # to be continued                         
#####################################################################################

# Set directory variables here
dir.nads <- "/ChIP/NADs"

NADs <- read.table(file.path(root,dir.nads, "NADsonly_mm39.csv"), sep= ",", header= F)
NADs <- NADs %>% transform( seqnames= V1, start = V2, end = V3)  %>% as_granges()

## Count overlaps of peaks within mouse for *selected genomic features*
nad.tbl <- mm39.tbl %>% mutate(non.nads_olap = count_overlaps(., mm39.tbl[!mm39.tbl %over% NADs,]),
                               nads_olap = count_overlaps(., mm39.tbl[mm39.tbl %over% NADs,]),
                               pol1_olap = count_overlaps(., pol1.bed),
                               pol2_olap = count_overlaps(., pol2.bed))

nad.tbl <- data.frame(nad.tbl)
nad.tbl$non.nads_olap<- ifelse(nad.tbl$non.nads_olap >1 , 1, nad.tbl$non.nads_olap)
nad.tbl$nads_olap<- ifelse(nad.tbl$nads_olap >1 , 1, nad.tbl$nads_olap)
nad.tbl$pol1_olap<- ifelse(nad.tbl$pol1_olap >1 , 1, nad.tbl$pol1_olap)
nad.tbl$pol2_olap<- ifelse(nad.tbl$pol2_olap >1 , 1, nad.tbl$pol2_olap)

## Prepare data for computing correlation
cor.tbl <- data.frame(nad.tbl)
rownames(cor.tbl) <- paste0(cor.tbl$seqnames, ": ", cor.tbl$start, "-", cor.tbl$end)
cor.tbl <- cor.tbl[,c(9:12)]

nrow(cor.tbl)
cor.tbl <- cor.tbl[ rowSums(cor.tbl) > 0, ]
nrow(cor.tbl)

## Non parametric Correlation test
M <- cor(cor.tbl, method="spearman") 

## heatmap of correlation coefficients
colors <- colorRampPalette(c("#009392", "white", "#e88471"))(64)

## Calculate correlation between pols across overlapping sections
p<-pheatmap(M, 
            color = colors,
            number_color = "black",
            fontsize = 16,
            fontsize_row = 16, 
            fontsize_col = 16,
            fontsize_number = 12,
            cellwidth = 50,
            cellheight = 50,
            cluster_cols= F,
            cluster_rows= F,
            display_numbers = T,
            labels_row = c("Within NADs", "Outside NADs", "Pol 1 Peaks", "Pol 2 Peaks"),
            labels_col = c("Within NADs", "Outside NADs", "Pol 1 Peaks", "Pol 2 Peaks"),
            main = "Spearman Correlation",
            annotation_legend = TRUE
)
p


```

#### Figure 6E: Spearman correlation

```{r, message=FALSE, warning=FALSE}

#####################################################################################
# Plot features of NAD Pol overlaps                       
#####################################################################################

## For plotting pol signal in and outside of NADs
pol1.anno <- GRanges(pol1.anno)
pol2.anno <- GRanges(pol2.anno)

## Code to either calculate average distance or average signal 
sum_peaks <- function(table, name){
  
  labels <- c("miRNA", "lncRNA","snoRNA", "snRNA","rRNA", "processed_pseudogene", 
              "transcribed_processed_pseudogene","retained_intron", "pseudogene", "protein_coding")
  
  ## Summmarize
  final.tbl <- table %>% group_by(tx_type) %>% 
    summarise(sum = log(sum(signalValue))) %>% 
    mutate(name = rep(name, times = length(unique(tx_type)), 
                      length.out = length(unique(tx_type)), each = 1))
  
  
  ## Filter
  final.tbl <- final.tbl[final.tbl$tx_type %in%labels ,]
  final.tbl <- final.tbl %>% arrange(factor(tx_type, levels = labels))
  final.tbl$factor = seq(1,nrow(final.tbl), by =1)
  
  return(final.tbl)
}

## Count overlaps of peaks within mouse for *selected genomic features*
pol1.tbl <- pol1.anno %>% mutate(non.nads_olap = count_overlaps(., pol1.anno[!pol1.anno %over% NADs,]),
                                 nads_olap = count_overlaps(., pol1.anno[pol1.anno %over% NADs,]),
                                 pol2_olap = count_overlaps(., pol2.bed))

pol2.tbl <- pol2.anno %>% mutate(non.nads_olap = count_overlaps(., pol2.anno[!pol2.anno %over% NADs,]),
                                 nads_olap = count_overlaps(., pol2.anno[pol2.anno %over% NADs,]),
                                 pol1_olap = count_overlaps(., pol1.bed))

pol1.tbl <- as.data.frame(pol1.tbl)
pol2.tbl <- as.data.frame(pol2.tbl)

## Get peaks where both pols overlap a feature
pol1.nad.tbl <- pol1.tbl[pol1.tbl$nads_olap >0,]
pol1.non.tbl <- pol1.tbl[pol1.tbl$non.nads_olap >0,]
pol2.nad.tbl <- pol2.tbl[nad.tbl$nads_olap >0,]
pol2.non.tbl <- pol2.tbl[nad.tbl$non.nads_olap >0,]
both.nad.tbl <- pol1.tbl[pol1.tbl$nads_olap >0 & pol1.tbl$pol2_olap >0,]

pol2.non.tbl <- sum_peaks(pol2.non.tbl, "Pol 2 non NADs")
pol2.tbl <- sum_peaks(pol2.nad.tbl, "Pol 2 NADs")
pol1.tbl <- sum_peaks(pol1.nad.tbl, "Pol 1 NADs")
pol1.tbl.2 <- sum_peaks(pol1.non.tbl, "Pol 1 non-NADs")
both.tbl <- sum_peaks(both.nad.tbl, "Both pols")

tbl <- rbind(pol2.tbl, pol1.tbl, pol1.tbl.2, both.tbl, pol2.non.tbl)

#colors <- c("#009392","#39b185","#9ccb86","#e9e29c","#eeb479","#e88471","#cf597e")
colors <- c("#009392","#9ccb86","#eeb479","#e88471","#cf597e")

labels <- c("miRNA", "lncRNA","snoRNA", "snRNA","rRNA", "processed_pseudogene", 
            "transcribed_processed_pseudogene","retained_intron", "pseudogene", "protein_coding")

## Plot individual timepoints as a barplot
barplot <- tbl %>% 
  mutate(tx_type= fct_reorder(tx_type, factor)) %>% 
  ggplot(aes(x=tx_type, y=sum, fill=name, colour = name), group_by(tx_type)) +
  scale_y_continuous(name="Log 10 Peak Signal", breaks = c(seq(0,14, by =1)) ,limits=c(0, 14))+
  geom_bar(stat="identity", width=0.75, position = "dodge", alpha = 1) +
  theme(axis.text.x = element_text(angle = 45,  hjust=1))+
  scale_color_manual(values=c("black","black","black","black","black"))+
  scale_fill_manual(values=colors) + theme_classic()+
  scale_x_discrete(limits = labels, labels = c("miRNA", "lncRNA","snoRNA", "snRNA","rRNA", "Processed Pseudogene", 
                                               "Transcribed Processed Pseudogene","Retained Intron", "Pseudogene", "Protein Coding"), expand = c(0.09, 0.09))+
  guides(fill = guide_legend(title = "Transcript Type")) 

barplot

## Code to either calculate average distance or average signal 
sum_peaks <- function(table, name){
  
  labels <- c("miRNA", "lncRNA","snoRNA", "snRNA","rRNA", "processed_pseudogene", 
              "transcribed_processed_pseudogene","retained_intron", "pseudogene", "protein_coding")
  
  ## Summmarize
  final.tbl <- table %>% group_by(tx_type) %>% 
    mutate(distanceToTSS = ifelse(distanceToTSS == 0 , 1, abs(distanceToTSS))) %>% 
    summarise(sum = mean(distanceToTSS),.groups = 'keep') %>% 
    mutate(name = rep(name, times = length(unique(tx_type)), 
                      length.out = length(unique(tx_type)), each = 1))
  
  
  ## Filter
  final.tbl <- final.tbl[final.tbl$tx_type %in%labels ,]
  final.tbl <- final.tbl %>% arrange(factor(tx_type, levels = labels))
  final.tbl$factor = seq(1,nrow(final.tbl), by =1)
  
  return(final.tbl)
}

## Count overlaps of peaks within mouse for *selected genomic features*
pol1.tbl <- pol1.anno %>% mutate(non.nads_olap = count_overlaps(., pol1.anno[!pol1.anno %over% NADs,]),
                                 nads_olap = count_overlaps(., pol1.anno[pol1.anno %over% NADs,]),
                                 pol2_olap = count_overlaps(., pol2.bed))

pol2.tbl <- pol2.anno %>% mutate(non.nads_olap = count_overlaps(., pol2.anno[!pol2.anno %over% NADs,]),
                                 nads_olap = count_overlaps(., pol2.anno[pol2.anno %over% NADs,]),
                                 pol1_olap = count_overlaps(., pol1.bed))

pol1.tbl <- as.data.frame(pol1.tbl)
pol2.tbl <- as.data.frame(pol2.tbl)

## Get peaks where both pols overlap a feature
pol1.nad.tbl <- pol1.tbl[pol1.tbl$nads_olap >0,]
pol1.non.tbl <- pol1.tbl[pol1.tbl$non.nads_olap >0,]
pol2.nad.tbl <- pol2.tbl[nad.tbl$nads_olap >0,]
pol2.non.tbl <- pol2.tbl[nad.tbl$non.nads_olap >0,]
both.nad.tbl <- pol1.tbl[pol1.tbl$nads_olap >0 & pol1.tbl$pol2_olap >0,]

pol2.non.tbl <- sum_peaks(pol2.non.tbl, "Pol 2 non NADs")
pol2.tbl <- sum_peaks(pol2.nad.tbl, "Pol 2 NADs")
pol1.tbl <- sum_peaks(pol1.nad.tbl, "Pol 1 NADs")
pol1.tbl.2 <- sum_peaks(pol1.non.tbl, "Pol 1 non-NADs")
both.tbl <- sum_peaks(both.nad.tbl, "Both pols")

tbl <- rbind(pol2.tbl, pol1.tbl, pol1.tbl.2, both.tbl, pol2.non.tbl)

## Plot individual timepoints as a barplot
barplot <- tbl %>% 
  mutate(tx_type= fct_reorder(tx_type, factor)) %>% 
  ggplot(aes(x=tx_type, y=sum, fill=name, colour = name), group_by(tx_type)) +
  scale_y_continuous(name="Distance to TSS (BP)", breaks = c(seq(0,70000, by =5000)) ,limits=c(0, 70000))+
  geom_bar(stat="identity", width=0.75, position = "dodge", alpha = 1) +
  theme(axis.text.x = element_text(angle = 45,  hjust=1))+
  scale_color_manual(values=c("black","black","black","black","black"))+
  scale_fill_manual(values=colors) + theme_classic()+
  scale_x_discrete(limits = labels, labels = c("miRNA", "lncRNA","snoRNA", "snRNA","rRNA", "Processed Pseudogene", 
                                               "Transcribed Processed Pseudogene","Retained Intron", "Pseudogene", "Protein Coding"), expand = c(0.09, 0.09))+
  guides(fill = guide_legend(title = "Transcript Type")) 

barplot

utils::sessionInfo()

```